// @contentstack/core/dist/esm/src/lib/contentstack-core.js
import _ from 'lodash';
import { serialize } from './param-serializer.js';
import axios, { getAdapter } from 'axios';
import { ERROR_MESSAGES } from './error-messages.js';
export function httpClient(options) {
    const defaultConfig = {
        insecure: false,
        retryOnError: true,
        headers: {},
        basePath: '',
        proxy: false,
        httpAgent: false,
        httpsAgent: false,
        timeout: 30000,
        logHandler: (level, data) => {
            if (level === 'error') {
                if (data) {
                    const title = [data.name, data.message].filter((a) => a).join(' - ');
                    console.error(ERROR_MESSAGES.CONSOLE.ERROR_WITH_TITLE(title));
                }
                return;
            }
            if (data !== undefined) {
                console.log(ERROR_MESSAGES.CONSOLE.LEVEL_WITH_DATA(level, data));
            }
        },
        retryCondition: (error) => {
            if (error.response && error.response.status === 429) {
                return true;
            }
            return false;
        },
        versioningStrategy: 'path',
    };
    const config = Object.assign(Object.assign({}, defaultConfig), _.cloneDeep(options));
    if (config.apiKey && config.headers) {
        config.headers.api_key = config.apiKey;
    }
    if (config.accessToken && config.headers) {
        config.headers.access_token = config.accessToken;
    }
    const protocol = config.insecure ? 'http' : 'https';
    const hostname = config.defaultHostname;
    const port = config.port || 443;
    const version = config.version || 'v3';
    const baseURL = config.endpoint || `${protocol}://${hostname}:${port}${config.basePath}/${version}`;
    const instance = axios.create(Object.assign(Object.assign({ 
        // Axios
        baseURL, adapter: getAdapter(axios.defaults.adapter) }, config), { paramsSerializer: {
            serialize,
        } }));
    instance.httpClientParams = options;
    if (config.onError) {
        instance.interceptors.response.use((response) => response, config.onError);
    }
    return instance;
}
