// @contentstack/delivery-sdk/dist/modern/stack/contentstack.cjs
'use strict';

var core = require('@contentstack/core');
var index = require('../cache/index.cjs');
var stack$1 = require('./stack.cjs');
var types = require('../common/types.cjs');
var errorMessages = require('../common/error-messages.cjs');
var utils$1 = require('../common/utils.cjs');
var utils = require('@contentstack/utils');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var utils__namespace = /*#__PURE__*/_interopNamespaceDefault(utils);

let version = "{{VERSION}}";
function stack(config) {
  const DEFAULT_HOST = utils$1.getHostforRegion(config.region || types.Region.US, config.host);
  let defaultConfig = {
    defaultHostname: DEFAULT_HOST,
    headers: {},
    params: {},
    live_preview: {},
    port: config.port,
    ...config
  };
  config.host = defaultConfig.defaultHostname;
  if (config.apiKey) {
    defaultConfig.headers.api_key = config.apiKey;
  } else {
    throw new Error("API key for Stack is required.");
  }
  if (config.deliveryToken) {
    defaultConfig.headers.access_token = config.deliveryToken;
  } else {
    throw new Error("Delivery token for Stack is required.");
  }
  if (config.environment) {
    defaultConfig.params.environment = config.environment;
  } else {
    throw new Error("Environment for Stack is required");
  }
  if (config.locale) {
    defaultConfig.params.locale = config.locale;
  }
  if (config.live_preview) {
    if (utils$1.isBrowser()) {
      const params = new URL(document.location.toString()).searchParams;
      if (params.has("live_preview")) {
        config.live_preview.live_preview = params.get("live_preview") || config.live_preview.live_preview;
      }
      if (params.has("release_id")) {
        defaultConfig.headers["release_id"] = params.get("release_id");
      }
      if (params.has("preview_timestamp")) {
        defaultConfig.headers["preview_timestamp"] = params.get("preview_timestamp");
      }
    }
  }
  if (config.branch) {
    defaultConfig.headers.branch = config.branch;
  }
  if (config.early_access) {
    defaultConfig.headers["x-header-ea"] = config.early_access.join(",");
  }
  defaultConfig.headers["X-User-Agent"] = "contentstack-delivery-typescript/" + version;
  const client = core.httpClient(defaultConfig);
  if (config.logHandler) client.defaults.logHandler = config.logHandler;
  if (config.cacheOptions && config.cacheOptions.policy !== types.Policy.IGNORE_CACHE) {
    if (!config.cacheOptions.persistenceStore) {
      throw new Error(errorMessages.ErrorMessages.MISSING_PERSISTENCE_STORE);
    }
    const defaultAdapter = client.defaults.adapter;
    client.defaults.adapter = (adapterConfig) => {
      return new Promise(async (resolve, reject) => {
        if (config.cacheOptions)
          await index.handleRequest(config.cacheOptions, config.apiKey, defaultAdapter, resolve, reject, adapterConfig);
      });
    };
  }
  if (config.debug) {
    client.interceptors.request.use((requestConfig) => {
      config.logHandler("info", {
        type: "request",
        method: requestConfig.method?.toUpperCase(),
        url: requestConfig.url,
        headers: requestConfig.headers,
        params: requestConfig.params,
        timestamp: (/* @__PURE__ */ new Date()).toISOString()
      });
      return requestConfig;
    });
    client.interceptors.response.use(
      (response) => {
        const level = getLogLevelFromStatus(response.status);
        config.logHandler(level, {
          type: "response",
          status: response.status,
          statusText: response.statusText,
          url: response.config?.url,
          method: response.config?.method?.toUpperCase(),
          headers: response.headers,
          data: response.data,
          timestamp: (/* @__PURE__ */ new Date()).toISOString()
        });
        return response;
      },
      (error) => {
        const status = error.response?.status || 0;
        const level = getLogLevelFromStatus(status);
        config.logHandler(level, {
          type: "response_error",
          status,
          statusText: error.response?.statusText || error.message,
          url: error.config?.url,
          method: error.config?.method?.toUpperCase(),
          error: error.message,
          timestamp: (/* @__PURE__ */ new Date()).toISOString()
        });
        throw error;
      }
    );
  }
  function getLogLevelFromStatus(status) {
    if (status >= 200 && status < 300) {
      return "info";
    } else if (status >= 300 && status < 400) {
      return "warn";
    } else if (status >= 400) {
      return "error";
    } else {
      return "debug";
    }
  }
  const errorHandler = (error) => {
    return core.retryResponseErrorHandler(error, config, client);
  };
  client.interceptors.request.use(core.retryRequestHandler);
  client.interceptors.response.use(core.retryResponseHandler, errorHandler);
  if (config.plugins) {
    client.interceptors.request.use((reqConfig) => {
      if (config && config.plugins)
        config.plugins.forEach((pluginInstance) => {
          reqConfig = pluginInstance.onRequest(reqConfig);
        });
      return reqConfig;
    });
    client.interceptors.response.use((response) => {
      if (config && config.plugins)
        config.plugins.forEach((pluginInstance) => {
          response = pluginInstance.onResponse(response.request, response, response.data);
        });
      return response;
    });
  }
  return new stack$1.Stack(client, config);
}

exports.Utils = utils__namespace;
exports.stack = stack;
