// @contentstack/delivery-sdk/dist/modern/stack/stack.cjs
'use strict';

var core = require('@contentstack/core');
var asset = require('../assets/asset.cjs');
var contenttypeQuery = require('../query/contenttype-query.cjs');
var assetQuery = require('../query/asset-query.cjs');
var taxonomyQuery = require('../query/taxonomy-query.cjs');
var globalFieldQuery = require('../query/global-field-query.cjs');
var index = require('../content-type/index.cjs');
var synchronization = require('../sync/synchronization.cjs');
var index$1 = require('../global-field/index.cjs');

class Stack {
  config;
  _client;
  constructor(client, config) {
    this._client = client;
    this.config = config;
    this._client.stackConfig = this.config;
  }
  asset(uid) {
    if (uid) return new asset.Asset(this._client, uid);
    return new assetQuery.AssetQuery(this._client);
  }
  contentType(uid) {
    if (uid) return new index.ContentType(this._client, uid);
    return new contenttypeQuery.ContentTypeQuery(this._client);
  }
  /**
     * @method Taxonomy
     * @memberof Stack
     * @description Sets the url to /taxonomies/entries. Pass a query to fetch entries with taxonomies
     *
     * @returns {TaxonomyQuery}
     * @example
     * import contentstack from '@contentstack/delivery-sdk'
     *
     * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
  
     * const taxonomy = stack.taxonomy() // For taxonomy query object
     */
  taxonomy() {
    return new taxonomyQuery.TaxonomyQuery(this._client);
  }
  globalField(uid) {
    if (uid) return new index$1.GlobalField(this._client, uid);
    return new globalFieldQuery.GlobalFieldQuery(this._client);
  }
  /**
   * @method setLocale
   * @memberof Stack
   * @description Sets the locale of the API server
   * @param {String} locale - valid locale e.g. fr-fr
   * @return {Stack}
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * stack.setLocale('en-155');
   */
  setLocale(locale) {
    this.config.locale = locale;
    this._client.defaults.params.locale = locale;
  }
  /**
   * @method sync
   * @memberof Stack
   * @description Syncs your Contentstack data with your app and ensures that the data is always up-to-date by providing delta updates
   * @param {object} params - params is an object that supports 'locale', 'start_date', 'content_type_uid', and 'type' queries.
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * await stack.sync()        // For initializing sync
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * await stack.sync({ 'locale': 'en-us'})     //For initializing sync with entries of a specific locale
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * await stack.sync({ 'start_date': '2018-10-22'})    //For initializing sync with entries published after a specific date
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * await stack.sync({ 'content_type_uid': 'session'})   //For initializing sync with entries of a specific content type
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * await stack.sync({ 'type': 'entry_published'})
   * //Use the type parameter to get a specific type of content. Supports'asset_published',
   * // 'entry_published', 'asset_unpublished', 'entry_unpublished', 'asset_deleted', 'entry_deleted', 'content_type_deleted'.
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * await stack.sync({'pagination_token': '<page_tkn>'})    // For fetching the next batch of entries using pagination token
   * @example
   * import contentstack from '@contentstack/delivery-sdk'
   *
   * const stack = contentstack.stack({ apiKey: "apiKey", deliveryToken: "deliveryToken", environment: "environment" });
   * await stack.sync({'sync_token': '<sync_tkn>'})    // For performing subsequent sync after initial sync
   * @instance
   */
  async sync(params = {}, recursive = false) {
    return await synchronization.synchronization(this._client, params, recursive);
  }
  livePreviewQuery(query) {
    if (this.config.live_preview) {
      let livePreviewParams = { ...this.config.live_preview };
      if (query.live_preview) {
        livePreviewParams = {
          ...livePreviewParams,
          live_preview: query.live_preview,
          contentTypeUid: query.contentTypeUid || query.content_type_uid,
          entryUid: query.entryUid || query.entry_uid,
          preview_timestamp: query.preview_timestamp || "",
          include_applied_variants: query.include_applied_variants || false
        };
      } else {
        livePreviewParams = {
          ...livePreviewParams,
          live_preview: "",
          contentTypeUid: "",
          entryUid: "",
          preview_timestamp: "",
          include_applied_variants: false
        };
      }
      this._client.stackConfig.live_preview = livePreviewParams;
    }
    if (query.hasOwnProperty("release_id")) {
      this._client.defaults.headers["release_id"] = query.release_id;
    } else {
      delete this._client.defaults.headers["release_id"];
    }
    if (query.hasOwnProperty("preview_timestamp")) {
      this._client.defaults.headers["preview_timestamp"] = query.preview_timestamp;
    } else {
      delete this._client.defaults.headers["preview_timestamp"];
    }
  }
  getClient() {
    return this._client;
  }
  async getLastActivities() {
    try {
      const result = await core.getData(this._client, "/content_types", {
        params: {
          only_last_activity: true,
          environment: this.config.environment
        }
      });
      return result;
    } catch (error) {
      throw new Error("Error fetching last activities");
    }
  }
  /**
   * @method setPort
   * @memberof Stack
   * @description Sets the port of the host
   * @param {Number} port - Port Number
   * @return {Stack}
   * @instance
   * */
  setPort(port) {
    if (typeof port === "number") this.config.port = port;
    return this;
  }
  /**
   * @method setDebug
   * @memberof Stack
   * @description Sets the debug option
   * @param {boolean} debug - Debug value
   * @return {Stack}
   * @instance
   * */
  setDebug(debug) {
    if (typeof debug === "boolean") this.config.debug = debug;
    return this;
  }
}

exports.Stack = Stack;
