// @contentstack/delivery-sdk/dist/modern/common/types.js
var Region = /* @__PURE__ */ ((Region2) => {
  Region2["US"] = "us";
  Region2["EU"] = "eu";
  Region2["AU"] = "au";
  Region2["AZURE_NA"] = "azure-na";
  Region2["AZURE_EU"] = "azure-eu";
  Region2["GCP_NA"] = "gcp-na";
  Region2["GCP_EU"] = "gcp-eu";
  return Region2;
})(Region || {});
var Policy = /* @__PURE__ */ ((Policy2) => {
  Policy2["IGNORE_CACHE"] = "IGNORE_CACHE";
  Policy2["CACHE_THEN_NETWORK"] = "CACHE_THEN_NETWORK";
  Policy2["CACHE_ELSE_NETWORK"] = "CACHE_ELSE_NETWORK";
  Policy2["NETWORK_ELSE_CACHE"] = "NETWORK_ELSE_CACHE";
  return Policy2;
})(Policy || {});
var PublishType = /* @__PURE__ */ ((PublishType2) => {
  PublishType2["ENTRY_PUBLISHED"] = "entry_published";
  PublishType2["ENTRY_UNPUBLISHED"] = "entry_unpublished";
  PublishType2["ENTRY_DELETED"] = "entry_deleted";
  PublishType2["ASSET_PUBLISHED"] = "asset_published";
  PublishType2["ASSET_UNPUBLISHED"] = "asset_unpublished";
  PublishType2["ASSET_DELETED"] = "asset_deleted";
  PublishType2["CONTENT_TYPE_DELETED"] = "content_type_deleted";
  return PublishType2;
})(PublishType || {});
var Format = /* @__PURE__ */ ((Format2) => {
  Format2["GIF"] = "gif";
  Format2["PNG"] = "png";
  Format2["JPG"] = "jpg";
  Format2["PJPG"] = "pjpg";
  Format2["WEBP"] = "webp";
  Format2["WEBPLL"] = "webpll";
  Format2["WEBPLY"] = "webply";
  return Format2;
})(Format || {});
var CropBy = /* @__PURE__ */ ((CropBy2) => {
  CropBy2["DEFAULT"] = "default";
  CropBy2["ASPECTRATIO"] = "aspectRatio";
  CropBy2["REGION"] = "region";
  CropBy2["OFFSET"] = "offset";
  return CropBy2;
})(CropBy || {});
var FitBy = /* @__PURE__ */ ((FitBy2) => {
  FitBy2["BOUNDS"] = "bounds";
  FitBy2["CROP"] = "crop";
  return FitBy2;
})(FitBy || {});
var Orientation = /* @__PURE__ */ ((Orientation2) => {
  Orientation2[Orientation2["DEFAULT"] = 1] = "DEFAULT";
  Orientation2[Orientation2["FLIP_HORIZONTAL"] = 2] = "FLIP_HORIZONTAL";
  Orientation2[Orientation2["FLIP_HORIZONTAL_VERTICAL"] = 3] = "FLIP_HORIZONTAL_VERTICAL";
  Orientation2[Orientation2["FLIP_VERTICAL"] = 4] = "FLIP_VERTICAL";
  Orientation2[Orientation2["FLIP_HORIZONTAL_LEFT"] = 5] = "FLIP_HORIZONTAL_LEFT";
  Orientation2[Orientation2["RIGHT"] = 6] = "RIGHT";
  Orientation2[Orientation2["FLIP_HORIZONTAL_RIGHT"] = 7] = "FLIP_HORIZONTAL_RIGHT";
  Orientation2[Orientation2["LEFT"] = 8] = "LEFT";
  return Orientation2;
})(Orientation || {});
var OverlayAlign = /* @__PURE__ */ ((OverlayAlign2) => {
  OverlayAlign2["TOP"] = "top";
  OverlayAlign2["BOTTOM"] = "bottom";
  OverlayAlign2["LEFT"] = "left";
  OverlayAlign2["RIGHT"] = "right";
  OverlayAlign2["MIDDLE"] = "middle";
  OverlayAlign2["CENTER"] = "center";
  return OverlayAlign2;
})(OverlayAlign || {});
var OverlayRepeat = /* @__PURE__ */ ((OverlayRepeat2) => {
  OverlayRepeat2["X"] = "x";
  OverlayRepeat2["Y"] = "y";
  OverlayRepeat2["BOTH"] = "both";
  return OverlayRepeat2;
})(OverlayRepeat || {});
var ResizeFilter = /* @__PURE__ */ ((ResizeFilter2) => {
  ResizeFilter2["NEAREST"] = "nearest";
  ResizeFilter2["BILINEAR"] = "bilinear";
  ResizeFilter2["BICUBIC"] = "bicubic";
  ResizeFilter2["LANCZOS2"] = "lanczos2";
  ResizeFilter2["LANCZOS3"] = "lanczos3";
  return ResizeFilter2;
})(ResizeFilter || {});
var CanvasBy = /* @__PURE__ */ ((CanvasBy2) => {
  CanvasBy2["DEFAULT"] = "default";
  CanvasBy2["ASPECTRATIO"] = "aspectRatio";
  CanvasBy2["REGION"] = "region";
  CanvasBy2["OFFSET"] = "offset";
  return CanvasBy2;
})(CanvasBy || {});
var QueryOperation = /* @__PURE__ */ ((QueryOperation2) => {
  QueryOperation2["EQUALS"] = "";
  QueryOperation2["NOT_EQUALS"] = "$ne";
  QueryOperation2["INCLUDES"] = "$in";
  QueryOperation2["EXCLUDES"] = "$nin";
  QueryOperation2["IS_LESS_THAN"] = "$lt";
  QueryOperation2["IS_LESS_THAN_OR_EQUAL"] = "$lte";
  QueryOperation2["IS_GREATER_THAN"] = "$gt";
  QueryOperation2["IS_GREATER_THAN_OR_EQUAL"] = "$gte";
  QueryOperation2["EXISTS"] = "$exists";
  QueryOperation2["MATCHES"] = "$regex";
  return QueryOperation2;
})(QueryOperation || {});
var TaxonomyQueryOperation = /* @__PURE__ */ ((TaxonomyQueryOperation2) => {
  TaxonomyQueryOperation2["ABOVE"] = "$above";
  TaxonomyQueryOperation2["BELOW"] = "$below";
  TaxonomyQueryOperation2["EQ_ABOVE"] = "$eq_above";
  TaxonomyQueryOperation2["EQ_BELOW"] = "$eq_below";
  return TaxonomyQueryOperation2;
})(TaxonomyQueryOperation || {});
var QueryOperator = /* @__PURE__ */ ((QueryOperator2) => {
  QueryOperator2["AND"] = "$and";
  QueryOperator2["OR"] = "$or";
  return QueryOperator2;
})(QueryOperator || {});

export { CanvasBy, CropBy, FitBy, Format, Orientation, OverlayAlign, OverlayRepeat, Policy, PublishType, QueryOperation, QueryOperator, Region, ResizeFilter, TaxonomyQueryOperation };
