// @contentstack/core/dist/cjs/src/lib/contentstack-core.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.httpClient = void 0;
const tslib_1 = require("tslib");
const lodash_1 = tslib_1.__importDefault(require("lodash"));
const param_serializer_1 = require("./param-serializer");
const axios_1 = tslib_1.__importStar(require("axios"));
const error_messages_1 = require("./error-messages");
function httpClient(options) {
    const defaultConfig = {
        insecure: false,
        retryOnError: true,
        headers: {},
        basePath: '',
        proxy: false,
        httpAgent: false,
        httpsAgent: false,
        timeout: 30000,
        logHandler: (level, data) => {
            if (level === 'error') {
                if (data) {
                    const title = [data.name, data.message].filter((a) => a).join(' - ');
                    console.error(error_messages_1.ERROR_MESSAGES.CONSOLE.ERROR_WITH_TITLE(title));
                }
                return;
            }
            if (data !== undefined) {
                console.log(error_messages_1.ERROR_MESSAGES.CONSOLE.LEVEL_WITH_DATA(level, data));
            }
        },
        retryCondition: (error) => {
            if (error.response && error.response.status === 429) {
                return true;
            }
            return false;
        },
        versioningStrategy: 'path',
    };
    const config = Object.assign(Object.assign({}, defaultConfig), lodash_1.default.cloneDeep(options));
    if (config.apiKey && config.headers) {
        config.headers.api_key = config.apiKey;
    }
    if (config.accessToken && config.headers) {
        config.headers.access_token = config.accessToken;
    }
    const protocol = config.insecure ? 'http' : 'https';
    const hostname = config.defaultHostname;
    const port = config.port || 443;
    const version = config.version || 'v3';
    const baseURL = config.endpoint || `${protocol}://${hostname}:${port}${config.basePath}/${version}`;
    const instance = axios_1.default.create(Object.assign(Object.assign({ 
        // Axios
        baseURL, adapter: (0, axios_1.getAdapter)(axios_1.default.defaults.adapter) }, config), { paramsSerializer: {
            serialize: param_serializer_1.serialize,
        } }));
    instance.httpClientParams = options;
    if (config.onError) {
        instance.interceptors.response.use((response) => response, config.onError);
    }
    return instance;
}
exports.httpClient = httpClient;
