// @contentstack/core/dist/cjs/src/lib/contentstack-error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.error = void 0;
function error(errorResponse) {
    const config = errorResponse.config;
    const response = errorResponse.response;
    if (!config || !response) {
        const errorObj = new Error();
        Object.assign(errorObj, errorResponse);
        errorObj.message = JSON.stringify(errorResponse);
        throw errorObj;
    }
    const data = response.data;
    const errorDetails = {
        status: response.status,
        statusText: response.statusText,
    };
    if (config.headers && config.headers.authtoken) {
        const token = `...${config.headers.authtoken.substr(-5)}`;
        config.headers.authtoken = token;
    }
    if (config.headers && config.headers.authorization) {
        const token = `...${config.headers.authorization.substr(-5)}`;
        config.headers.authorization = token;
    }
    errorDetails.request = {
        url: config.url,
        method: config.method,
        data: config.data,
        headers: config.headers,
    };
    if (data) {
        errorDetails.errorMessage = data.error_message || data.message || '';
        errorDetails.errorCode = data.error_code || 0;
        errorDetails.errors = data.errors || {};
        errorDetails.error = data.error || '';
    }
    const errorObj = new Error();
    Object.assign(errorObj, errorDetails);
    errorObj.message = JSON.stringify(errorDetails);
    return errorObj;
}
exports.error = error;
