// @contentstack/delivery-sdk/dist/legacy/stack/contentstack.js
import { httpClient, retryRequestHandler, retryResponseHandler, retryResponseErrorHandler } from '@contentstack/core';
import { handleRequest } from '../cache/index.js';
import { Stack } from './stack.js';
import { Region, Policy } from '../common/types.js';
import { ErrorMessages } from '../common/error-messages.js';
import { getHostforRegion, isBrowser } from '../common/utils.js';
import * as utils from '@contentstack/utils';
export { utils as Utils };

let version = "{{VERSION}}";
function stack(config) {
  const DEFAULT_HOST = getHostforRegion(config.region || Region.US, config.host);
  let defaultConfig = {
    defaultHostname: DEFAULT_HOST,
    headers: {},
    params: {},
    live_preview: {},
    port: config.port,
    ...config
  };
  config.host = defaultConfig.defaultHostname;
  if (config.apiKey) {
    defaultConfig.headers.api_key = config.apiKey;
  } else {
    throw new Error("API key for Stack is required.");
  }
  if (config.deliveryToken) {
    defaultConfig.headers.access_token = config.deliveryToken;
  } else {
    throw new Error("Delivery token for Stack is required.");
  }
  if (config.environment) {
    defaultConfig.params.environment = config.environment;
  } else {
    throw new Error("Environment for Stack is required");
  }
  if (config.locale) {
    defaultConfig.params.locale = config.locale;
  }
  if (config.live_preview) {
    if (isBrowser()) {
      const params = new URL(document.location.toString()).searchParams;
      if (params.has("live_preview")) {
        config.live_preview.live_preview = params.get("live_preview") || config.live_preview.live_preview;
      }
      if (params.has("release_id")) {
        defaultConfig.headers["release_id"] = params.get("release_id");
      }
      if (params.has("preview_timestamp")) {
        defaultConfig.headers["preview_timestamp"] = params.get("preview_timestamp");
      }
    }
  }
  if (config.branch) {
    defaultConfig.headers.branch = config.branch;
  }
  if (config.early_access) {
    defaultConfig.headers["x-header-ea"] = config.early_access.join(",");
  }
  defaultConfig.headers["X-User-Agent"] = "contentstack-delivery-typescript/" + version;
  const client = httpClient(defaultConfig);
  if (config.logHandler) client.defaults.logHandler = config.logHandler;
  if (config.cacheOptions && config.cacheOptions.policy !== Policy.IGNORE_CACHE) {
    if (!config.cacheOptions.persistenceStore) {
      throw new Error(ErrorMessages.MISSING_PERSISTENCE_STORE);
    }
    const defaultAdapter = client.defaults.adapter;
    client.defaults.adapter = (adapterConfig) => {
      return new Promise(async (resolve, reject) => {
        if (config.cacheOptions)
          await handleRequest(config.cacheOptions, config.apiKey, defaultAdapter, resolve, reject, adapterConfig);
      });
    };
  }
  if (config.debug) {
    client.interceptors.request.use((requestConfig) => {
      config.logHandler("info", {
        type: "request",
        method: requestConfig.method?.toUpperCase(),
        url: requestConfig.url,
        headers: requestConfig.headers,
        params: requestConfig.params,
        timestamp: (/* @__PURE__ */ new Date()).toISOString()
      });
      return requestConfig;
    });
    client.interceptors.response.use(
      (response) => {
        const level = getLogLevelFromStatus(response.status);
        config.logHandler(level, {
          type: "response",
          status: response.status,
          statusText: response.statusText,
          url: response.config?.url,
          method: response.config?.method?.toUpperCase(),
          headers: response.headers,
          data: response.data,
          timestamp: (/* @__PURE__ */ new Date()).toISOString()
        });
        return response;
      },
      (error) => {
        const status = error.response?.status || 0;
        const level = getLogLevelFromStatus(status);
        config.logHandler(level, {
          type: "response_error",
          status,
          statusText: error.response?.statusText || error.message,
          url: error.config?.url,
          method: error.config?.method?.toUpperCase(),
          error: error.message,
          timestamp: (/* @__PURE__ */ new Date()).toISOString()
        });
        throw error;
      }
    );
  }
  function getLogLevelFromStatus(status) {
    if (status >= 200 && status < 300) {
      return "info";
    } else if (status >= 300 && status < 400) {
      return "warn";
    } else if (status >= 400) {
      return "error";
    } else {
      return "debug";
    }
  }
  const errorHandler = (error) => {
    return retryResponseErrorHandler(error, config, client);
  };
  client.interceptors.request.use(retryRequestHandler);
  client.interceptors.response.use(retryResponseHandler, errorHandler);
  if (config.plugins) {
    client.interceptors.request.use((reqConfig) => {
      if (config && config.plugins)
        config.plugins.forEach((pluginInstance) => {
          reqConfig = pluginInstance.onRequest(reqConfig);
        });
      return reqConfig;
    });
    client.interceptors.response.use((response) => {
      if (config && config.plugins)
        config.plugins.forEach((pluginInstance) => {
          response = pluginInstance.onResponse(response.request, response, response.data);
        });
      return response;
    });
  }
  return new Stack(client, config);
}

export { stack };
