// @contentstack/core/dist/cjs/src/lib/retryPolicy/delivery-sdk-handlers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.calculateRateLimitDelay = exports.retryResponseErrorHandler = exports.retryResponseHandler = exports.retryRequestHandler = exports.getRetryDelay = void 0;
const tslib_1 = require("tslib");
const error_messages_1 = require("../error-messages");
const defaultConfig = {
    maxRequests: 5,
    retryLimit: 5,
    retryDelay: 300,
};
const DEFAULT_RETRY_DELAY_MS = 300;
/**
 * Resolve retry delay from customBackoff, configured retryDelay, or default.
 */
const getRetryDelay = (config) => {
    var _a;
    if ((_a = config.retryDelayOptions) === null || _a === void 0 ? void 0 : _a.customBackoff) {
        return config.retryDelayOptions.customBackoff();
    }
    const retryDelay = config.retryDelay;
    if (retryDelay) {
        return retryDelay;
    }
    return DEFAULT_RETRY_DELAY_MS;
};
exports.getRetryDelay = getRetryDelay;
const retryRequestHandler = (req) => {
    req.retryCount = req.retryCount || 1;
    return req;
};
exports.retryRequestHandler = retryRequestHandler;
const retryResponseHandler = (response) => response;
exports.retryResponseHandler = retryResponseHandler;
const retryResponseErrorHandler = (error, config, axiosInstance) => {
    try {
        let retryCount = error.config.retryCount;
        config = Object.assign(Object.assign({}, defaultConfig), config);
        if (!error.config.retryOnError || retryCount > config.retryLimit) {
            throw error;
        }
        const response = error.response;
        if (!response) {
            if (config.retryCondition && config.retryCondition(error)) {
                retryCount++;
                return retry(error, config, retryCount, (0, exports.getRetryDelay)(config), axiosInstance);
            }
            if (error.code === 'ECONNABORTED') {
                const customError = {
                    error_message: error_messages_1.ERROR_MESSAGES.RETRY.TIMEOUT_EXCEEDED(config.timeout),
                    error_code: error_messages_1.ERROR_MESSAGES.ERROR_CODES.TIMEOUT,
                    errors: null,
                };
                throw customError; // Throw customError object
            }
            throw error;
        }
        const rateLimitRemaining = response.headers['x-ratelimit-remaining'];
        // Handle rate limit exhaustion with retry logic
        if (rateLimitRemaining !== undefined && parseInt(rateLimitRemaining) <= 0) {
            retryCount++;
            if (retryCount >= config.retryLimit) {
                return Promise.reject(error.response.data);
            }
            error.config.retryCount = retryCount;
            // Calculate delay for rate limit reset
            const rateLimitResetDelay = (0, exports.calculateRateLimitDelay)(response.headers);
            return new Promise((resolve, reject) => {
                setTimeout(() => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
                    try {
                        const retryResponse = yield axiosInstance(error.config);
                        resolve(retryResponse);
                    }
                    catch (retryError) {
                        reject(retryError);
                    }
                }), rateLimitResetDelay);
            });
        }
        if (response.status == 429 || response.status == 401) {
            retryCount++;
            if (retryCount >= config.retryLimit) {
                if (error.response && error.response.data) {
                    return Promise.reject(error.response.data);
                }
                return Promise.reject(error);
            }
            error.config.retryCount = retryCount;
            // Apply configured delay for retries
            return new Promise((resolve, reject) => {
                setTimeout(() => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
                    try {
                        const retryResponse = yield axiosInstance(error.config);
                        resolve(retryResponse);
                    }
                    catch (retryError) {
                        reject(retryError);
                    }
                }), (0, exports.getRetryDelay)(config));
            });
        }
        if (config.retryCondition && config.retryCondition(error)) {
            retryCount++;
            return retry(error, config, retryCount, (0, exports.getRetryDelay)(config), axiosInstance);
        }
        throw error;
    }
    catch (err) {
        throw err;
    }
};
exports.retryResponseErrorHandler = retryResponseErrorHandler;
const retry = (error, config, retryCount, retryDelay, axiosInstance) => {
    if (retryCount > config.retryLimit) {
        return Promise.reject(error);
    }
    const delayTime = retryDelay || (0, exports.getRetryDelay)(config);
    error.config.retryCount = retryCount;
    return new Promise(function (resolve, reject) {
        return setTimeout(function () {
            return tslib_1.__awaiter(this, void 0, void 0, function* () {
                try {
                    const retryResponse = yield axiosInstance(error.config);
                    resolve(retryResponse);
                }
                catch (retryError) {
                    reject(retryError);
                }
            });
        }, delayTime);
    });
};
/**
 * Calculate delay time for rate limit reset based on response headers
 * @param headers - Response headers from the API
 * @returns Delay time in milliseconds
 */
const calculateRateLimitDelay = (headers) => {
    // Check for retry-after header (in seconds)
    const retryAfter = headers['retry-after'];
    if (retryAfter) {
        return parseInt(retryAfter) * 1000; // Convert to milliseconds
    }
    // Check for x-ratelimit-reset header (Unix timestamp)
    const rateLimitReset = headers['x-ratelimit-reset'];
    if (rateLimitReset) {
        const resetTime = parseInt(rateLimitReset) * 1000; // Convert to milliseconds
        const currentTime = Date.now();
        const delay = resetTime - currentTime;
        // Ensure we have a positive delay, add a small buffer
        return Math.max(delay + 1000, 1000); // At least 1 second delay
    }
    // Check for x-ratelimit-reset-time header (ISO string)
    const rateLimitResetTime = headers['x-ratelimit-reset-time'];
    if (rateLimitResetTime) {
        const resetTime = new Date(rateLimitResetTime).getTime();
        const currentTime = Date.now();
        const delay = resetTime - currentTime;
        // Ensure we have a positive delay, add a small buffer
        return Math.max(delay + 1000, 1000); // At least 1 second delay
    }
    // Default fallback delay (1 second) if no rate limit reset info is available
    return 1000;
};
exports.calculateRateLimitDelay = calculateRateLimitDelay;
