// @contentstack/delivery-sdk/dist/legacy/cache/index.js
import { Policy } from '../common/types.js';
import { ErrorMessages } from '../common/error-messages.js';

function extractEntryUidFromUrl(config) {
  if (!config.url) return null;
  const entryUrlPattern = /\/content_types\/[^\/]+\/entries\/([^\/\?]+)/;
  const match = config.url.match(entryUrlPattern);
  return match ? match[1] : null;
}
function generateEnhancedCacheKey(originalKey, contentTypeUid, entryUid) {
  let cacheKey = originalKey;
  if (contentTypeUid) {
    cacheKey = `${contentTypeUid}_${cacheKey}`;
  }
  if (entryUid) {
    cacheKey = `${cacheKey}_entry_${entryUid}`;
  }
  return cacheKey;
}
async function handleRequest(cacheOptions, apiKey, defaultAdapter, resolve, reject, config) {
  if (!cacheOptions.persistenceStore) {
    throw new Error(ErrorMessages.MISSING_PERSISTENCE_STORE);
  }
  const cacheStore = cacheOptions.persistenceStore;
  const entryUid = config.entryUid || extractEntryUidFromUrl(config);
  const enhancedCacheKey = generateEnhancedCacheKey(apiKey, config.contentTypeUid, entryUid);
  switch (cacheOptions.policy) {
    case Policy.NETWORK_ELSE_CACHE: {
      const apiResponse = await defaultAdapter(config);
      if (apiResponse.data) {
        cacheStore.setItem(enhancedCacheKey, JSON.parse(apiResponse.data), config.contentTypeUid, cacheOptions.maxAge);
        return resolve({ data: JSON.parse(apiResponse.data) });
      } else {
        const cacheResponse = cacheStore.getItem(enhancedCacheKey, config.contentTypeUid);
        if (cacheResponse)
          return resolve({
            data: cacheResponse,
            status: 200,
            statusText: "OK",
            headers: {},
            config: {}
          });
      }
      return reject(apiResponse);
    }
    case Policy.CACHE_THEN_NETWORK: {
      const cacheResponse = cacheStore.getItem(enhancedCacheKey, config.contentTypeUid);
      if (cacheResponse)
        return resolve({
          data: cacheResponse,
          status: 200,
          statusText: "OK",
          headers: {},
          config: {}
        });
      const apiResponse = await defaultAdapter(config);
      if (apiResponse.data) {
        cacheStore.setItem(enhancedCacheKey, JSON.parse(apiResponse.data), config.contentTypeUid, cacheOptions.maxAge);
        return resolve({ data: JSON.parse(apiResponse.data) });
      } else {
        return reject(apiResponse);
      }
    }
    case Policy.CACHE_ELSE_NETWORK: {
      const cacheResponse = cacheStore.getItem(enhancedCacheKey, config.contentTypeUid);
      if (cacheResponse)
        return resolve({
          data: cacheResponse,
          status: 200,
          statusText: "OK",
          headers: {},
          config: {}
        });
      else {
        const apiResponse = await defaultAdapter(config);
        if (apiResponse.data) {
          cacheStore.setItem(enhancedCacheKey, JSON.parse(apiResponse.data), config.contentTypeUid, cacheOptions.maxAge);
          return resolve({ data: JSON.parse(apiResponse.data) });
        } else {
          return reject(apiResponse);
        }
      }
    }
  }
}

export { handleRequest };
